Userland/Libraries/LibJS/Runtime/ArrayConstructor.cpp

```diff
@@ -95,13 +95,29 @@ JS_DEFINE_NATIVE_FUNCTION(ArrayConstructor::from)
 
     auto* array = Array::create(global_object);
 
+    Function* map_fn = nullptr;
+    if (vm.argument_count() >= 2) {
+        auto callback = vm.argument(1);
+        if (!callback.is_function()) {
+            vm.throw_exception<TypeError>(global_object, ErrorType::NotAFunction, callback.to_string_without_side_effects());
+            return {};
+        }
+        map_fn = &callback.as_function();
+    }
+
     // Array.from() lets you create Arrays from:
     if (auto size = object->indexed_properties().array_like_size()) {
         // * array-like objects (objects with a length property and indexed elements)
         MarkedValueList elements(vm.heap());
         elements.ensure_capacity(size);
         for (size_t i = 0; i < size; ++i) {
-            elements.append(object->get(i));
+            if (map_fn) {
+                auto map_fn_result = vm.call(*map_fn, value, object->get(i));
```

Comment:
Missing exception check, when `object->get()` throws, the subsequent `call()` will blow up. Needs to be split into separate steps.